src/Algebra/Graph/AdjacencyIntMap/Algorithm.hs

```diff
@@ -213,32 +243,86 @@ dfs vs = concatMap flatten . dfsForestFrom vs
 -- 'isSubgraphOf' ('vertices' $ reachable x y) y == True
 -- @
 reachable :: Int -> AdjacencyIntMap -> [Int]
-reachable x = concat . bfs [x]
+reachable x = dfs [x]
+
+type Cycle = NonEmpty
+data NodeState = Entered | Exited
+data S = S { table :: IntMap.IntMap Int
+           , entry :: IntMap.IntMap NodeState
+           , order :: [Int] }
+
+topSort' :: (MonadState S m, MonadCont m)
+         => AdjacencyIntMap -> m (Either (Cycle Int) [Int])
+topSort' g = callCC $ \cyclic ->
+  do let vertices = map fst $ IntMap.toDescList $ adjacencyIntMap g
+         adjacent = IntSet.toDescList . flip postIntSet g
+         dfsRoot x = nodeState x >>= \case
+           Nothing -> enterRoot x >> dfs x >> exit x
+           _       -> return ()
+         dfs x = forM_ (adjacent x) $ \y ->
+                   nodeState y >>= \case
+                     Nothing      -> enter x y >> dfs y >> exit y
+                     Just Exited  -> return ()
+                     Just Entered -> cyclic . Left . retrace x y =<< gets table
+     forM_ vertices dfsRoot
+     Right <$> gets order
+  where
+    nodeState v = gets (IntMap.lookup v . entry)
+    enter u v = modify' (\(S m n vs) -> S (IntMap.insert v u m)
+                                          (IntMap.insert v Entered n)
+                                          vs)
+    enterRoot v = modify' (\(S m n vs) -> S m (IntMap.insert v Entered n) vs)
+    exit v = modify' (\(S m n vs) -> S m (IntMap.alter (fmap leave) v n) (v:vs))
+      where leave = \case
+              Entered -> Exited
+              Exited  -> error "Internal error: dfs search order violated"
+    retrace curr head table = aux (curr :| []) where
+      aux xs@(curr :| _)
+        | head == curr = xs
+        | otherwise = case table IntMap.! curr of
+            prev -> aux (prev <| xs)
```

Comment:
I'm a bit confused about why this branch is implemented this way. Could we do the following instead?

```
aux (table IntMap.! curr <| xs)
```

Reply:
Good point! This was an artifact from earlier when the pattern matching was needed